client/components/app-header/index.jsx

```diff
@@ -7,43 +7,99 @@ import {
-    </svg>
-  );
-
-  let qrIcon = (
-    <svg fill="#FFFF4C" width="20pt" height="20pt" version="1.1" viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg">
-      <path d="m24.309 78.371h18.98zm4.7461-14.234h9.4883v9.4883h-9.4883zm23.727-9.4883h4.7461v18.98h-4.7461zm-33.215 28.469h28.469v-28.469h-28.469zm4.7422-4.7461v-18.98h18.98v18.98zm28.473 0h9.4883v4.7461h-9.4922zm23.723-14.234h4.7461v4.7461h-4.7461zm-9.4883-9.4883h14.234v4.7461l-14.234-0.003906zm-4.7461-23.727h9.4883v9.4883h-9.4883zm18.98-9.4883h-28.469v28.469h28.469zm-4.7461 23.723h-18.98v-18.98h18.98l0.003906 18.98zm-56.938 4.7461h28.469v-28.469h-28.469zm4.7422-23.727h18.98v18.98h-18.98zm52.195 52.195h-9.4883v4.7461h14.234v-9.4883h-4.7461zm-14.234-14.234v9.4883h9.4883v-4.7461h-4.7461l0.003906-4.7422zm-33.215-33.215h9.4883v9.4883h-9.4883z" />
-    </svg>
-  );
-
-  return (
-    <header className='header'>
-      <Appbar className='mui--appbar-line-height'>
-        <div className="mui-container-fluid">
-          <a className="sidedrawer-toggle mui--visible-xs-inline-block mui--visible-sm-inline-block js-show-sidedrawer" onClick={doLeftToggle}>{appLogo}</a>
-          <a className="sidedrawer-toggle mui--hidden-xs mui--hidden-sm js-hide-sidedrawer" onClick={doLeftToggle}>{appLogo}</a>
-          <span className="mui--text-title mui--inline-block">
-            <Link to="/">Frontend Grocer</Link>
-          </span>
-          <a className="appbar-icon cart-toggle mui--pull-right" onClick={doRightToggle}>
-            {cartIcon}
-            {numItemsInCart > 0 ? <span className="cart-item-count-badge">{numItemsInCart}</span> : ''}
-          </a>
-          <a className="appbar-icon qr-reader mui--pull-right" >{qrIcon}</a>
-        </div>
-      </Appbar>
-    </header>
-  )
-};
+class AppHeader extends Component{
+  constructor(props) {
+    super(props);
+    this.onQrButtonClicked = this.onQrButtonClicked.bind(this);
+    this.onQrImageSelected = this.onQrImageSelected.bind(this);
+    this.state = {
+      inputClassName: `input-field-${Math.round(Math.random() * 100000).toString(16)}`
+    }
+  }
+
+  onQrButtonClicked() {
+    let fileInput = document.getElementsByClassName(this.state.inputClassName)[0];
+    fileInput.click();
+  }
+
+  onQrImageSelected(evt) {
+    //TODO handle errors properly
+    return new Promise((resolve/*, reject*/) => {
```

Comment:
This is ugly as hell, but the TL;DR is
* use a file input to grab the file
* use FileReader to encode it into a base64 url string
* Use that url to paint the image onto a canvas
* Grab the image buffer from the canvas

AFAIK this is the only way we can do image analysis in a worker, due to limited ability to instantiate these kinds of objects (i.e., Image, Canvas, etc...) in the worker scope.

Goes without saying this is code that students shouldn't worry about -- it doesn't need to change as we move QR analysis from main-thread to worker-thread